src/toolkitAPI.c

```diff
@@ -33,12 +35,84 @@ int  stats_getLinkStat(int index, SM_LinkStats *linkStats);
 int  stats_getPumpStat(int index, SM_PumpStats *pumpStats);
 TSubcatchStats *stats_getSubcatchStat(int index);
 
+
 // Utilty Function Declarations
 double* newDoubleArray(int n);
 
+
 //-----------------------------------------------------------------------------
 //  Extended API Functions
 //-----------------------------------------------------------------------------
+
+int DLLEXPORT  swmm_run_cb(const char* f1, const char* f2, const char* f3,
+    void (*callback) (double *))
+//
+//  Input:   f1 = name of input file
+//           f2 = name of report file
+//           f3 = name of binary output file
+//  Output:  returns error code
+//  Purpose: runs a SWMM simulation.
+//
+{
+    clock_t check = 0;
+    double progress, elapsedTime = 0.0;
+
+
+    // --- initialize flags
+    IsOpenFlag = FALSE;
+    IsStartedFlag = FALSE;
+    SaveResultsFlag = TRUE;
+
+
+    // --- open the files & read input data
+    ErrorCode = 0;
+    swmm_open(f1, f2, f3);
+
+    // --- run the simulation if input data OK
+    if ( !ErrorCode )
+    {
+        // --- initialize values
+        swmm_start(TRUE);
+
+        // --- execute each time step until elapsed time is re-set to 0
+        if ( !ErrorCode )
+        {
+            do
+            {
+                swmm_step(&elapsedTime);
+
+                // --- callback with progress approximately twice a second
+                if ( (callback != NULL) && (clock() - check) > CLOCKS_PER_SEC )
+                {
+                    progress = NewRoutingTime / TotalDuration;
+                    callback(&progress);
+                    check = clock();
+                }
```

Comment:
@michaeltryby, Is there a reason you are using a clock here instead of checking the fraction of the simulation period that has been completed? It seems to me that using a timer will lead to multiple and potentially unecessary callbacks for long duration simulation depending on the size of CLOCKS_PER_SEC. An approach I have seen used is to check if a certain fraction of progress has been made before calling the callback function.

Reply:
The previous implementation would updating progress each time an hour of simulation time had elapsed. This works because progress was being displayed to the user in terms of elapsed simulation time. This new approach differs in that it displays percent of simulation completed and remaining wall time. A timer is used to update these metrics on an interval that is frequent enough but not too frequent. The current timer settings update progress approximately twice a second or at the end of the timestep which ever is longer. In practice this seems to work well.

Reply:
What is the use case for this API function? Can you provide an example?